component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -296,8 +297,22 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
         if (userToken.equals(contextToken)) {
             context.setProperty(EmailOTPAuthenticatorConstants.OTP_TOKEN, "");
             context.setProperty(EmailOTPAuthenticatorConstants.EMAILOTP_ACCESS_TOKEN, "");
-            String emailFromProfile = context.getProperty(EmailOTPAuthenticatorConstants.RECEIVER_EMAIL).toString();
-            context.setSubject(AuthenticatedUser.createFederateAuthenticatedUserFromSubjectIdentifier(emailFromProfile));
+            Map<Integer, StepConfig> stepConfigMap = context.getSequenceConfig().getStepMap();
+            AuthenticatedUser authenticatedUser = null;
+            for (StepConfig stepConfig : stepConfigMap.values()) {
+                authenticatedUser = stepConfig.getAuthenticatedUser();
+                if (authenticatedUser != null && stepConfig.isSubjectAttributeStep()) {
+                    break;
+                }
+            }
+            if (authenticatedUser == null) {
```

Comment:
Lets simplify the logic here inside the fr loop. We can assign the value to authenticatedUser, if it is the subject step and the authenticated user is not null

Reply:
Fixed: 538a79627d56a2893bad088599fe197d4e42cc03